File: articles/calculus-for-systems-modeling/articles/introduction-to-partial-differential-equations/r/pde_diffusion_grid_audit.R
initialize_field <- function(grid_points) {
  field <- rep(0, grid_points)
  center <- ceiling(grid_points / 2)
  field[[center]] <- 1
  field
}

diffusion_step <- function(field, stability_ratio) {
  updated <- field
  for (i in 2:(length(field) - 1)) {
    updated[[i]] <- field[[i]] + stability_ratio * (field[[i + 1]] - 2 * field[[i]] + field[[i - 1]])
  }
  updated[[1]] <- 0
  updated[[length(updated)]] <- 0
  updated
}

simulate_diffusion <- function(grid_points, diffusivity, dx, dt, steps) {
  stability_ratio <- diffusivity * dt / (dx ^ 2)
  field <- initialize_field(grid_points)
  records <- list()
  for (step in 0:steps) {
    time <- step * dt
    records[[length(records) + 1]] <- data.frame(
      step = step,
      time = time,
      center_value = field[[ceiling(grid_points / 2)]],
      total_mass = sum(field) * dx,
      max_value = max(field),
      min_value = min(field),
      stability_ratio = stability_ratio,
      warning = "Explicit diffusion schemes require stability checks; boundary and grid assumptions shape results."
    )
    field <- diffusion_step(field, stability_ratio)
  }
  do.call(rbind, records)
}

results <- simulate_diffusion(
  grid_points = 51,
  diffusivity = 0.1,
  dx = 1,
  dt = 0.25,
  steps = 100
)

summary_table <- data.frame(
  grid_points = 51,
  diffusivity = 0.1,
  dx = 1,
  dt = 0.25,
  stability_ratio = 0.1 * 0.25 / (1 ^ 2),
  interpretation = "The grid audit records how a concentrated initial field spreads under diffusion-like dynamics."
)

dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_pde_diffusion_grid_audit.csv", row.names = FALSE)
write.csv(summary_table, "outputs/tables/r_pde_diffusion_summary.csv", row.names = FALSE)

print(head(results))
print(summary_table)

File: articles/calculus-for-systems-modeling/articles/introduction-to-partial-differential-equations/fortran/pde_diffusion_grid_audit.f90
program pde_diffusion_grid_audit
  implicit none
  integer, parameter :: grid_points = 51, steps = 100
  integer :: step, i
  real(8) :: diffusivity, dx, dt, ratio, total_mass, max_value, min_value
  real(8), dimension(grid_points) :: field, updated

  diffusivity = 0.1d0
  dx = 1.0d0
  dt = 0.25d0
  ratio = diffusivity * dt / (dx * dx)
  field = 0.0d0
  field((grid_points + 1) / 2) = 1.0d0

  print '(A)', 'step time center_value total_mass max_value min_value stability_ratio'
  do step = 0, steps
    total_mass = sum(field) * dx
    max_value = maxval(field)
    min_value = minval(field)
    print '(I6,6F14.6)', step, dble(step)*dt, field((grid_points + 1) / 2), total_mass, max_value, min_value, ratio

    updated = field
    do i = 2, grid_points - 1
      updated(i) = field(i) + ratio * (field(i + 1) - 2.0d0 * field(i) + field(i - 1))
    end do
    updated(1) = 0.0d0
    updated(grid_points) = 0.0d0
    field = updated
  end do
end program
